rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/mqtt/TbMqttNodeTest.java

```diff
@@ -15,26 +15,298 @@
+        mqttNodeConfig.setCredentials(new CertPemCredentials());
+
+        mockConnectClient(mqttNode);
+        given(promiseMock.get(anyLong(), any(TimeUnit.class))).willReturn(resultMock);
+        given(resultMock.isSuccess()).willReturn(true);
+
+        assertThatNoException().isThrownBy(() -> mqttNode.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(mqttNodeConfig))));
+    }
+
+    @Test
+    public void givenFailedByTimeoutConnectResult_whenInit_thenThrowsException() throws ExecutionException, InterruptedException, TimeoutException {
+        mqttNodeConfig.setHost("localhost");
+        mqttNodeConfig.setClientId("bfrbTESTmfkr23");
+        mqttNodeConfig.setCredentials(new CertPemCredentials());
+
+        mockConnectClient(mqttNode);
+        given(promiseMock.get(anyLong(), any(TimeUnit.class))).willThrow(new TimeoutException("Failed to connect"));
+
+        assertThatThrownBy(() -> mqttNode.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(mqttNodeConfig))))
+                .isInstanceOf(TbNodeException.class)
+                .hasMessage("java.lang.RuntimeException: Failed to connect to MQTT broker at localhost:1883.")
+                .extracting(e -> ((TbNodeException) e).isUnrecoverable())
+                .isEqualTo(false);
+    }
+
+    @Test
+    public void givenFailedConnectResult_whenInit_thenThrowsException() throws ExecutionException, InterruptedException, TimeoutException {
+        mqttNodeConfig.setHost("localhost");
+        mqttNodeConfig.setClientId("bfrbTESTmfkr23");
+        mqttNodeConfig.setAppendClientIdSuffix(true);
+        mqttNodeConfig.setCredentials(new CertPemCredentials());
+
+        mockConnectClient(mqttNode);
+        given(promiseMock.get(anyLong(), any(TimeUnit.class))).willReturn(resultMock);
+        given(resultMock.isSuccess()).willReturn(false);
+        given(resultMock.getReturnCode()).willReturn(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+
+        assertThatThrownBy(() -> mqttNode.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(mqttNodeConfig))))
+                .isInstanceOf(TbNodeException.class)
+                .hasMessage("java.lang.RuntimeException: Failed to connect to MQTT broker at localhost:1883. Result code is: CONNECTION_REFUSED_NOT_AUTHORIZED")
+                .extracting(e -> ((TbNodeException) e).isUnrecoverable())
+                .isEqualTo(false);
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    public void givenForceAckIsTrueAndTopicPatternAndIsRetainedMsgIsTrue_whenOnMsg_thenTellSuccess(String topicPattern, TbMsgMetaData metaData, String data) {
+        mqttNodeConfig.setRetainedMessage(true);
+        mqttNodeConfig.setTopicPattern(topicPattern);
+        ReflectionTestUtils.setField(mqttNode, "mqttNodeConfiguration", mqttNodeConfig);
+        ReflectionTestUtils.setField(mqttNode, "mqttClient", mqttClientMock);
+        ReflectionTestUtils.setField(mqttNode, "forceAck", true);
```

Comment:
to many reflection. Please rewrite tests without them. Our plan was to create a method for getting mqtt client that we will be able to mock and as a result this allow us always call init method when needed.

```
    protected MqttClient getClient(TbContext ctx, MqttClientConfig config) {
        return MqttClient.create(config, null, ctx.getExternalCallExecutor());
    }
```